File: iks-core/src/com/iks/cms/core/sql/Table.java
package com.iks.cms.core.sql;

import java.util.*;

/**
 * @author Igor Kaynov
 */
public class Table {
  private String name;
  private String alias;
  private Map< String, Column > columns = new HashMap<>();
  public Table( String name ) {
    this.name = name;
  }
  public Table( String name, String alias ) {
    this( name );
    this.alias = alias;
  }
  public String toString() {
    return alias != null ? SqlUtils.quote(name) + " as " + SqlUtils.quote( alias ) : SqlUtils.quote( name );
  }
  public String getAlias() {
    return alias;
  }
  public String getName() {
    return name;
  }
  public Column getColumn( String fieldName ) {
    return getColumn( fieldName, null );
  }
  public Column getColumn( String fieldName, String fieldAlias ) {
    Column column = columns.get( fieldName );
    if( column == null ) {
      column = new Column( this, fieldName, fieldAlias );
      columns.put( fieldName, column );
    }
    return column;
  }
  @Override
  public boolean equals(Object o) {
    if (o == null || !(o instanceof Table)) {
      return false;
    }
    Table otherTable = (Table) o;
    return Objects.equals(  name, otherTable.name);
  }
}

File: iks-core/src/com/iks/cms/core/sql/Table.kt
package com.iks.cms.core.sql

import java.util.*

/**
 * @author Igor Kaynov
 */
class Table(val name: String) {
  var alias: String? = null
  private val columns = HashMap<String, Column>()

  constructor(name: String, alias: String) : this(name) {
    this.alias = alias
  }

  override fun toString(): String {
    return if (alias != null) name.quote() + " as " + alias.quote() else name.quote()
  }

  @JvmOverloads fun getColumn(fieldName: String, fieldAlias: String? = null): Column {
    var column: Column? = columns[fieldName]
    if (column == null) {
      column = Column(this, fieldName, fieldAlias)
      columns.put(fieldName, column)
    }
    return column
  }

  override fun equals(other: Any?): Boolean {
    if (other == null || other !is Table) {
      return false
    }
    return name == other.name
  }

  companion object {
    private val counter = 0
  }
}
